### DISCUSSION TITLE
Key equality in ReduceByKey

### DISCUSSION THREAD

**Discussion by kevkrist** [NVIDIA/cccl#1407] [CUB] [answered]
Is there a way to use a custom type with a custom equality operator as the key type for ReduceByKey? Implementing `__device__ bool operator==(const CustomType& other)` on the type isn't doing the trick (failing to build). Is that expected?

I'm particularly interested in a key simply composed of a small sequence of primitive types. Looks like this might be possible if I use a thrust zip_iterator as the KeyIterator? Another possible solution, it seems, is to modify the interface so I can pass in a custom equality operator to dispatch?

**Comment by pauleonix** [comment]
`zip_iterator` is great in terms of giving coalesced memory access, but the `tuple` that is then used for comparison already has an `operator==` defined. Not sure if you can or should create a specialization of it.

I'm pretty sure that `operator==` is a valid customization point here, so you will need to add more information/code/compiler output.

**Reply by kevkrist** [reply, marked as answer]
It turns out I needed `__device__ bool operator==(const CustomType& other) const {...}` (I was missing the last `const`), which did the trick. Thank you! and sorry for wasting anyone's time.
